aws/chef-bootstrap.sh

```diff
@@ -0,0 +1,129 @@
+#!/bin/bash
+# Code-dot-org Chef bootstrap script.
+# One-liner to run:
+# aws s3 cp s3://cdo-dist/chef/bootstrap.sh - | sudo bash -s -- [options]
+#
+# Options:
+# -b [branch]
+# -n [node_name]
+# -r [run_list]
+# -v [chef_version]
+# -e [environment]
+# -z: Use local-mode Chef client.
+#
+# Prerequisites:
+# curl
+# awscli (pip install awscli)
+# AWS read permissions for s3://cdo-dist/ provided by environment (Instance IAM profile, ~/.aws/credentials, etc)
+# [local mode] branch-specific Chef cookbooks uploaded to S3
+
+# Set script defaults
+ENVIRONMENT=adhoc
+BRANCH=staging
+CHEF_VERSION=12.7.2
+RUN_LIST='recipe[cdo-apps]'
+NODE_NAME=$(hostname)
+
+# Parse options
+while getopts ":b:n:r:v:e:z" opt; do
+  case "${opt}" in
+    e)
+      ENVIRONMENT=${OPTARG}
+      ;;
+    b)
+      BRANCH=${OPTARG}
+      ;;
+    n)
+      NODE_NAME=${OPTARG}
+      ;;
+    r)
+      RUN_LIST=${OPTARG}
+      ;;
+    v)
+      CHEF_VERSION=${OPTARG}
+      ;;
+    z)
+      LOCAL_MODE=1
+      ;;
+    \?)
+      echo "Invalid option: -${OPTARG}" >&2
+      ;;
+  esac
+done
+
+# Redirect copy of stdout/stderr to a log file for later auditing.
+LOG=/var/log/chef-bootstrap.log
+exec >> >(tee -i ${LOG})
+exec 2>&1
+
+CHEF_CLIENT=/opt/chef/bin/chef-client
+# Ensure correct version of Chef is installed.
+if [ "$(${CHEF_CLIENT} -v)" != "Chef: ${CHEF_VERSION}" ]; then
+  curl -L https://omnitruck.chef.io/install.sh | bash -s -- -v ${CHEF_VERSION}
+else
+  echo "Chef ${CHEF_VERSION} is installed."
+fi
+${CHEF_CLIENT} -v
+
+mkdir -p /etc/chef
+CLIENT_RB=/etc/chef/client.rb
+cat <<RUBY > ${CLIENT_RB}
+log_level :info
+ssl_verify_mode :verify_peer
+node_name '${NODE_NAME}'
+environment '${ENVIRONMENT}'
+unless '${LOCAL_MODE}' == '1'
+  validation_client_name   'code-dot-org-validator'
+  chef_server_url          'https://api.chef.io/organizations/code-dot-org'
+end
+RUBY
+
+# Write default first-boot.json to be used by the chef-client command.
+# Existing file takes precedence.
+FIRST_BOOT=/etc/chef/first-boot.json
+if [ ! -f ${FIRST_BOOT} ] ; then
+  cat <<JSON > ${FIRST_BOOT}
+{
+  "omnibus_updater": {
+    "version": "${CHEF_VERSION}"
+  },
+  "cdo-repository": {
+    "branch": "${BRANCH}"
+  },
+  "run_list": ["${RUN_LIST}"]
+}
+JSON
+fi
+
+# Remove existing Chef Client key every time bootstrap is run.
+if [ -f /etc/chef/client.pem ] ; then
+    rm /etc/chef/client.pem
+fi
+
+if [ "${LOCAL_MODE}" = "1" ]; then
+  mkdir -p /opt/chef-zero/{cookbooks,environments}
+  # Install branch-specific Chef cookbooks from s3.
+  REPO_COOKBOOK_URL=s3://cdo-dist/chef/${BRANCH}.tar.gz
+  aws s3 cp ${REPO_COOKBOOK_URL} - | tar xz -C /opt/chef-zero
```

Comment:
According to the [documentation](http://docs.aws.amazon.com/cli/latest/topic/s3-faq.html#download), the command-line interface verifies checksums, and attempts up to 3 retries, then exits with a non-zero return code (in which case the piped `tar` command would also exit with a non-zero return code).

This should be sufficient to silently handle intermittent s3 operation failures (if it is not, I'd raise an issue in the quite-responsive [awscli](https://github.com/aws/aws-cli) project).

In the case of a more persistent (non-recoverable) failure, the whole bootstrap script should fail and exit with a non-zero return code. To ensure that this case would stop the bootstrap immediately I'll add [`set -o errexit`](http://pubs.opengroup.org/onlinepubs/009696799/utilities/set.html) to the top of this script so it exits immediately on any failure.